### PR TITLE
jit: Conv3d + BatchNorm3d fusion

### PR DESCRIPTION
**Pull request by XiaobingSuper**

### CODE REVIEW

File: test/quantization/test_quantize_script.py [hunk 1]
```diff
@@ -50,9 +50,9 @@ class TestQuantizeScriptJitPasses(QuantizationTestCase):
     """
     def test_foldbn_trivial(self):
         # Test trivial case
-        class TestModule(torch.nn.Module):
+        class TestModule2d(torch.nn.Module):
```
Comment (jerryzh168) [inline test/quantization/test_quantize_script.py:53]: can you follow: https://github.com/pytorch/pytorch/blob/master/test/quantization/test_quantize_script.py#L1771 and merge the two class into one?

File: test/quantization/test_quantize_script.py [hunk 2]
```diff
@@ -62,40 +62,53 @@ def forward(self, x):
...
-            FileCheck().check_count("prim::CallMethod[name=\"forward\"]", 2, exactly=True) \
-                .run(str(get_forward(scripted_or_traced._c).graph))
+        def _test_output(eager, x):
+            # Check that the transformation doesn't change numerics
+            for tracing_mode in [True, False]:
+                eager.eval()
+                if tracing_mode:
+                    scripted_or_traced = torch.jit.trace(eager, x)
+                else:
+                    scripted_or_traced = torch.jit.script(eager)
```
Comment (jerryzh168) [inline test/quantization/test_quantize_script.py:84]: we have a utility function: `get_script_module`

File: test/quantization/test_quantize_script.py [hunk 3]
```diff
@@ -62,40 +62,53 @@ def forward(self, x):
...
+                return x
 
-            # Check that in the original script module's forward we have two
-            # CallMethod nodes. One of them should be for conv.forward and the other
-            # for bn.forward.
-            FileCheck().check_count("prim::CallMethod[name=\"forward\"]", 2, exactly=True) \
-                .run(str(get_forward(scripted_or_traced._c).graph))
+        def _test_output(eager, x):
+            # Check that the transformation doesn't change numerics
+            for tracing_mode in [True, False]:
```
Comment (jerryzh168) [inline test/quantization/test_quantize_script.py:79]: nit: please rename `tracing_mode` to `tracing` to align with other tests

File: test/quantization/test_quantize_script.py [hunk 4]
```diff
@@ -62,40 +62,53 @@ def forward(self, x):
...
+                # Check that the transformation doesn't change numerics
+                self.assertEqual(eager(x), scripted_or_traced(x))
+
+        _test_output(TestModule2d(), torch.rand(1, 1, 6, 6))
+        _test_output(TestModule3d(), torch.rand(1, 1, 6, 6, 6))
 
     def test_foldbn_trivial_nobias(self):
         # Test trivial case
-        class TestModule(torch.nn.Module):
+        class TestModule2d(torch.nn.Module):
```
Comment (jerryzh168) [inline test/quantization/test_quantize_script.py:109]: same here, please merge the two classes into one class

File: test/quantization/test_quantize_script.py [hunk 5]
```diff
@@ -107,39 +120,54 @@ def forward(self, x):
...
+                # Check that the transformation doesn't change numerics
+                self.assertEqual(eager(x), scripted_or_traced(x))
+
+        _test_output(TestModule2d(), torch.rand(1, 1, 6, 6))
+        _test_output(TestModule3d(), torch.rand(1, 1, 6, 6, 6))
 
     def test_foldbn_in_submodule(self):
         # Test that we find Conv-BN patterns in submodules
-        class SubModule(torch.nn.Module):
+        class SubModule2d(torch.nn.Module):
```
Comment (jerryzh168) [inline test/quantization/test_quantize_script.py:168]: same here

File: test/quantization/test_quantize_script.py [hunk 6]
```diff
@@ -148,91 +176,133 @@ def forward(self, x):
...
-            self.assertEqual(eager(x), scripted_or_traced(x))
+        _test_output(TestModule2d(), torch.rand(1, 1, 10, 10))
+        _test_output(TestModule3d(), torch.rand(1, 1, 10, 10, 10))
 
     def test_foldbn_in_customConv2D(self):
-        # Make sure a custom Conv2D class is not folded
+        # Make sure a custom Conv2D or Conv3d class is not folded
         # as we do not know it does.
-        class CustomConv2D(torch.nn.Module):
+        class CustomConv(torch.nn.Module):
```
Comment (jerryzh168) [inline test/quantization/test_quantize_script.py:233]: wait, how is this a conv module?

Comment (jerryzh168) [reply test/quantization/test_quantize_script.py:233]: oh, nvm, this test can probably be removed, we will have more exact match for conv modules

File: torch/csrc/jit/passes/fold_conv_bn.cpp [hunk 7]
```diff
@@ -353,12 +346,32 @@ void FoldConvBatchNorm2dHelper::transform() {
...
-  addBiasForConv2dIfNone(m);
-  h.analyze(m);
-  h.transform();
+
+  addBiasForConvIfNone(m, "Conv2d");
+  addBiasForConvIfNone(m, "Conv3d");
+  // Conv2d + BatchNorm2d
+  const PatternInfo pattern2d = PatternInfo::parse_from_str(R"IR(
+graph(%self, %x):
+    %conv_submodule = match::module[name=".Conv2d"](%self)
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/fold_conv_bn.cpp:358]: we can now match more exactly, please take a look at https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/passes/quantization/insert_observers.cpp.html#507

File: torch/csrc/jit/passes/fold_conv_bn.cpp [hunk 8]
```diff
@@ -353,12 +346,32 @@ void FoldConvBatchNorm2dHelper::transform() {
...
+  // Conv3d + BatchNorm3d
+  const PatternInfo pattern3d = PatternInfo::parse_from_str(R"IR(
+graph(%self, %x):
+    %conv_submodule = match::module[name=".Conv3d"](%self)
+    %conv_out = prim::CallMethod[name="forward"](%conv_submodule, %x)
+    %bn_submodule = match::module[name=".BatchNorm3d"](%self)
+    %bn_out = prim::CallMethod[name="forward"](%bn_submodule, %conv_out)
+    return (%bn_out))IR");
+  h2.analyze(m, pattern3d);
+  h2.transform();
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/fold_conv_bn.cpp:374]: also, I think it's better to do this in a loop, since we might have conv1d+bn1d as well

**Review by jerryzh168 — CHANGES_REQUESTED** [review]
Thanks for adding this! Please see some comments inline.

File: torch/csrc/jit/passes/fold_conv_bn.cpp [hunk 9]
```diff
@@ -353,12 +347,39 @@ void FoldConvBatchNorm2dHelper::transform() {
...
+  // Conv2d + BatchNorm2d
+  const PatternInfo pattern2d = PatternInfo::parse_from_str(
+      R"(
+graph(%self, %x):
+    %conv_submodule = match::module[name=".Conv2d"](%self)
+    %conv_out = prim::CallMethod[name="forward"](%conv_submodule, %x)
+    %bn_submodule = match::module[name=".BatchNorm2d"](%self)
+    %bn_out = prim::CallMethod[name="forward"](%bn_submodule, %conv_out)
+    return (%bn_out))",
+      {is_conv2d_module, is_batchnorm2d_module});
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/fold_conv_bn.cpp:364]: you'll need to update the pattern as well, `is_conv2d_module` is checking the type for `%conv`.
You can take a look at https://github.com/pytorch/pytorch/pull/39490/files to see how the patterns are updated

Comment (jerryzh168) [reply torch/csrc/jit/passes/fold_conv_bn.cpp:364]: the patterns should be something like:
```
graph(%x, %conv, %batchnorm):
    %conv_out = prim::CallMethod[name="forward"](%conv, %x)
    %bn_out = prim::CallMethod[name="forward"](%batchnorm, %conv_out)
    return (%bn_out))",
```

File: torch/csrc/jit/passes/fold_conv_bn.cpp [hunk 10]
```diff
@@ -56,10 +57,9 @@ void replaceConv2dBiasWithGetAttr(Module& module) {
   }
 }
 
-void addBiasForConv2dIfNone(Module& module) {
+void addBiasForConvIfNone(Module& module, const std::string& pattern_name) {
   auto t = module.type()->expect<ClassType>();
   auto real_typename = t->name()->qualifiedName();
-  const std::string pattern_name("Conv2d");
   if (real_typename.size() >= pattern_name.size() &&
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/fold_conv_bn.cpp:63]: these probably should be updated with a regex

File: torch/csrc/jit/passes/fold_conv_bn.cpp (hunk 9 above)
Comment (XiaobingSuper) [reply torch/csrc/jit/passes/fold_conv_bn.cpp:362]: Changed.

**Review by jerryzh168 — APPROVED** [review]
Looks good, thanks!

**Comment by jerryzh168** [comment]
@XiaobingSuper could you rebase, there is some conflict

**Comment by XiaobingSuper** [comment]
rebased.
